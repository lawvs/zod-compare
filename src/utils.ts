import { z, type ZodType } from "zod";

export interface IsSameTypeOptions {
  deep: true;
  /**
   * Ignore all specific validations like min, max, length, email etc.
   *
   * You still can use `interceptor` to validate the type manually.
   */
  ignoreValidations: true;
  ignoreOptional: boolean;
  ignoreNullable: boolean;
  ignoreReadOnly: false;
  ignoreBranded: boolean;
  ignoreIntersectionsOrder: false;
  /**
   * A function that provides custom logic for comparing two ZodType instances.
   *
   * If the function returns `true` or `false`, the result will be used as the comparison result.
   * Otherwise, the default comparison logic will be used.
   *
   */
  interceptor: (
    a: ZodType,
    b: ZodType,
    options: IsSameTypeOptions,
  ) => boolean | void;
}

export interface IsCompatibleTypeOptions extends IsSameTypeOptions {
  ignoreOptional: false;
  ignoreNullable: false;
}

export const DEFAULT_COMPARE_TYPE_OPTIONS = {
  deep: true,
  ignoreOptional: false,
  ignoreNullable: false,
  ignoreReadOnly: false,
  ignoreBranded: false,
  ignoreValidations: true,
  ignoreIntersectionsOrder: false,
  interceptor: () => {},
} as const satisfies IsSameTypeOptions;

// See ZodFirstPartyTypeKind
export const isPrimitiveType = (a: ZodType): boolean => {
  return (
    a instanceof z.ZodString ||
    a instanceof z.ZodNumber ||
    a instanceof z.ZodNaN ||
    a instanceof z.ZodBigInt ||
    a instanceof z.ZodBoolean ||
    a instanceof z.ZodDate ||
    a instanceof z.ZodSymbol ||
    a instanceof z.ZodUndefined ||
    a instanceof z.ZodNull ||
    a instanceof z.ZodAny ||
    a instanceof z.ZodUnknown ||
    a instanceof z.ZodNever ||
    a instanceof z.ZodVoid
  );
};
