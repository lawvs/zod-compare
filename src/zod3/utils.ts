import { z, type ZodType } from "zod/v3";

// See ZodFirstPartyTypeKind
export const isPrimitiveType = (a: ZodType): boolean => {
  return (
    a instanceof z.ZodString ||
    a instanceof z.ZodNumber ||
    a instanceof z.ZodNaN ||
    a instanceof z.ZodBigInt ||
    a instanceof z.ZodBoolean ||
    a instanceof z.ZodDate ||
    a instanceof z.ZodSymbol ||
    a instanceof z.ZodUndefined ||
    a instanceof z.ZodNull ||
    a instanceof z.ZodAny ||
    a instanceof z.ZodUnknown ||
    a instanceof z.ZodNever ||
    a instanceof z.ZodVoid
  );
};

type Mutable<T> = { -readonly [k in keyof T]: T[k] };

export const flatUnwrapUnion = <
  T extends z.ZodUnionOptions = readonly [z.ZodTypeAny, ...z.ZodTypeAny[]],
>(
  t: z.ZodUnion<T>,
): Mutable<T> => {
  return t.options.flatMap((x) => {
    if (x instanceof z.ZodUnion) {
      return flatUnwrapUnion(x);
    }
    return x;
  }) as unknown as T;
};
