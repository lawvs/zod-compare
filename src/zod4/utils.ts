import type {
  $ZodAny,
  $ZodBigInt,
  $ZodBoolean,
  $ZodDate,
  $ZodNaN,
  $ZodNever,
  $ZodNull,
  $ZodNumber,
  $ZodString,
  $ZodSymbol,
  $ZodType,
  $ZodTypes,
  $ZodUndefined,
  $ZodUnion,
  $ZodUnknown,
  $ZodVoid,
  SomeType,
} from "zod/v4/core";

export const isZodType = (maybeSchema: unknown): maybeSchema is $ZodType => {
  if (typeof maybeSchema !== "object" || maybeSchema === null) {
    return false;
  }
  if (
    !("~standard" in maybeSchema) ||
    typeof maybeSchema["~standard"] !== "object"
  ) {
    return false;
  }
  return (
    (maybeSchema["~standard"] as $ZodType["~standard"])["vendor"] === "zod"
  );
};

const typesSet: Set<string> = new Set([
  "string",
  "number",
  "boolean",
  "bigint",
  "symbol",
  "null",
  "undefined",
  "void",
  "never",
  "any",
  "unknown",
  "date",
  "object",
  "record",
  "file",
  "array",
  "tuple",
  "union",
  "intersection",
  "map",
  "set",
  "enum",
  "literal",
  "nullable",
  "optional",
  "nonoptional",
  "success",
  "transform",
  "default",
  "prefault",
  "catch",
  "nan",
  "pipe",
  "readonly",
  "template_literal",
  "promise",
  "lazy",
  "custom",
] satisfies $ZodTypes["_zod"]["def"]["type"][]);

export const isZodTypes = (schema: $ZodType): schema is $ZodTypes => {
  const type = schema._zod.def.type;
  return typesSet.has(type);
};

export const isSimpleType = (
  a: $ZodTypes,
): a is
  | $ZodString
  | $ZodNumber
  | $ZodBoolean
  | $ZodBigInt
  | $ZodSymbol
  | $ZodUndefined
  | $ZodNull
  | $ZodAny
  | $ZodUnknown
  | $ZodNever
  | $ZodVoid
  | $ZodDate
  | $ZodNaN => {
  const type = a._zod.def.type;
  return (
    type === "string" ||
    type === "number" ||
    type === "boolean" ||
    type === "bigint" ||
    type === "symbol" ||
    type === "undefined" ||
    type === "null" ||
    type === "any" ||
    type === "unknown" ||
    type === "never" ||
    type === "void" ||
    type === "date" ||
    type === "nan" ||
    type === "file"
  );
};

export const flatUnwrapUnion = <
  Options extends readonly SomeType[] = readonly $ZodType[],
>(
  unionType: $ZodUnion<Options>,
): Options => {
  return unionType._zod.def.options.flatMap((x) => {
    if (
      x._zod.def.type === "union" &&
      Array.isArray((x as $ZodUnion)._zod.def.options)
    ) {
      return flatUnwrapUnion(x as $ZodUnion);
    }
    return x;
  }) as unknown as Options;
};

export const compareSchemas = (
  a: $ZodTypes,
  b: $ZodTypes,
  compareZod: (a: $ZodTypes, b: $ZodTypes) => boolean,
): boolean => {
  const aDef = a._zod.def;
  const bDef = b._zod.def;
  for (const key in aDef) {
    if (!(key in bDef)) return false;
    const aValue = aDef[key as keyof typeof aDef];
    const bValue = bDef[key as keyof typeof bDef];

    if (typeof aValue !== typeof bValue) return false;
    if (aValue === null && bValue !== null) return false;
    if (
      typeof aValue === "string" ||
      typeof aValue === "number" ||
      typeof aValue === "boolean" ||
      typeof aValue === "symbol" ||
      typeof aValue === "bigint" ||
      typeof aValue === "undefined"
    ) {
      if (aValue !== bValue) return false;
      continue;
    }
    if (isZodType(aValue)) {
      if (!isZodType(bValue)) return false;
      const isEqual = compareZod(
        aValue as unknown as $ZodTypes,
        bValue as unknown as $ZodTypes,
      );
      if (!isEqual) return false;
      continue;
    }
    if (Array.isArray(aValue)) {
      if (!Array.isArray(bValue)) return false;
      if (aValue.length !== bValue.length) return false;
      for (let i = 0; i < aValue.length; i++) {
        const aItem = aValue[i];
        const bItem = bValue[i];
        if (isZodType(aItem)) {
          if (!isZodType(bItem)) return false;
          if (!compareZod(aItem as $ZodTypes, bItem as $ZodTypes)) return false;
          continue;
        }
        if (aItem !== bItem) {
          return false;
        }
      }
      continue;
    }
    if (typeof aValue === "object") {
      if (typeof bValue !== "object" || bValue === null) return false;
      const bShape = bValue;
      if (Object.keys(aValue).length !== Object.keys(bShape).length)
        return false;
      console.warn("Skipping object comparison", key, aValue, bValue);
      continue;
    }
    if (typeof aValue === "function") {
      // skip function comparison
      console.warn("Skipping function comparison", key, aValue, bValue);
      continue;
    }
  }
  return true;
};
