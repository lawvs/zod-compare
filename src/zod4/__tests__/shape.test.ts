import { beforeAll, describe, expect, test } from "vitest";
import { z } from "zod/v4";
import type { $ZodType } from "zod/v4/core";
import { z as zm } from "zod/v4/mini";
import { isZodType } from "../utils.ts";

describe("snapshot", () => {
  beforeAll(() => {
    const weakSet = new WeakSet();
    expect.addSnapshotSerializer({
      serialize(val: $ZodType, config, indentation, depth, refs, printer) {
        const normalizeZodVal = (val: $ZodType) => {
          return {
            ...("_zod" in val
              ? {
                  _zod: {
                    ...("def" in val._zod ? { def: val._zod["def"] } : {}),
                  },
                }
              : {}),
            ...("def" in val ? { def: val["def"] } : {}),
            "~standard": val["~standard"],
          };
        };
        const normalizedVal = normalizeZodVal(val);
        weakSet.add(normalizedVal);

        return printer(normalizedVal, config, indentation, depth, refs);
      },
      test(val) {
        if (weakSet.has(val)) {
          return false;
        }
        return isZodType(val);
        // return (
        //   val &&
        //   typeof val === "object" &&
        //   ("def" in val || "standard" in val || "_zod" in val)
        // );
      },
    });
  });

  test("should z.string() match snapshot", () => {
    expect(z.string()).toMatchSnapshot();
    expect(zm.string()).toMatchSnapshot();
  });

  test("should z.literal() match snapshot", () => {
    expect(z.literal("test")).toMatchSnapshot();
    expect(zm.literal("test")).toMatchSnapshot();
  });

  test("should z.object() match snapshot", () => {
    expect(
      z.object({ foo: z.string(), bar: z.number().nullable() }),
    ).toMatchSnapshot();
    expect(
      zm.object({ foo: zm.string(), bar: zm.nullable(zm.number()) }),
    ).toMatchSnapshot();
  });

  test("should z.array() match snapshot", () => {
    expect(z.array(z.string())).toMatchSnapshot();
    expect(zm.array(zm.string())).toMatchSnapshot();
  });

  test("should z.tuple() match snapshot", () => {
    expect(z.tuple([z.string(), z.number()], z.boolean())).toMatchSnapshot();
    expect(zm.tuple([zm.string(), zm.number()], z.boolean())).toMatchSnapshot();
  });

  test("should z.union() match snapshot", () => {
    expect(z.union([z.string(), z.number()])).toMatchSnapshot();
    expect(zm.union([zm.string(), zm.number()])).toMatchSnapshot();
  });

  test("should z.intersection() match snapshot", () => {
    expect(z.intersection(z.string(), z.number())).toMatchSnapshot();
    expect(zm.intersection(zm.string(), zm.number())).toMatchSnapshot();
  });

  test("should z.record() match snapshot", () => {
    expect(z.record(z.string(), z.number())).toMatchSnapshot();
    expect(zm.record(zm.string(), zm.number())).toMatchSnapshot();
  });

  test("should z.enum() match snapshot", () => {
    expect(z.enum(["foo", "bar", "baz"])).toMatchSnapshot();
    expect(zm.enum(["foo", "bar", "baz"])).toMatchSnapshot();
  });

  test("should z.function() match snapshot", () => {
    expect(
      z.function({
        input: [z.string()],
        output: z.number(),
      }),
    ).toMatchSnapshot();
    expect(
      zm.function({ input: [z.string()], output: z.number() }),
    ).toMatchSnapshot();
  });
});
