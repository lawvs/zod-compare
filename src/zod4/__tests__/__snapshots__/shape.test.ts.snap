// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`snapshot > should z.array() match snapshot 1`] = `
{
  "_zod": {
    "def": {
      "element": {
        "_zod": {
          "def": {
            "type": "string",
          },
        },
        "def": {
          "type": "string",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      "type": "array",
    },
  },
  "def": {
    "element": {
      "_zod": {
        "def": {
          "type": "string",
        },
      },
      "def": {
        "type": "string",
      },
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "type": "array",
  },
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`snapshot > should z.array() match snapshot 2`] = `
{
  "_zod": {
    "def": {
      "element": {
        "_zod": {
          "def": {
            "type": "string",
          },
        },
        "def": {
          "type": "string",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      "type": "array",
    },
  },
  "def": {
    "element": {
      "_zod": {
        "def": {
          "type": "string",
        },
      },
      "def": {
        "type": "string",
      },
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "type": "array",
  },
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`snapshot > should z.enum() match snapshot 1`] = `
{
  "_zod": {
    "def": {
      "entries": {
        "bar": "bar",
        "baz": "baz",
        "foo": "foo",
      },
      "type": "enum",
    },
  },
  "def": {
    "entries": {
      "bar": "bar",
      "baz": "baz",
      "foo": "foo",
    },
    "type": "enum",
  },
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`snapshot > should z.enum() match snapshot 2`] = `
{
  "_zod": {
    "def": {
      "entries": {
        "bar": "bar",
        "baz": "baz",
        "foo": "foo",
      },
      "type": "enum",
    },
  },
  "def": {
    "entries": {
      "bar": "bar",
      "baz": "baz",
      "foo": "foo",
    },
    "type": "enum",
  },
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`snapshot > should z.function() match snapshot 1`] = `
$ZodFunction {
  "_def": {
    "input": {
      "_zod": {
        "def": {
          "items": [
            {
              "_zod": {
                "def": {
                  "type": "string",
                },
              },
              "def": {
                "type": "string",
              },
              "~standard": {
                "validate": [Function],
                "vendor": "zod",
                "version": 1,
              },
            },
          ],
          "rest": null,
          "type": "tuple",
        },
      },
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "output": {
      "_zod": {
        "def": {
          "checks": [],
          "type": "number",
        },
      },
      "def": {
        "checks": [],
        "type": "number",
      },
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "type": "function",
  },
  "def": {
    "input": {
      "_zod": {
        "def": {
          "items": [
            {
              "_zod": {
                "def": {
                  "type": "string",
                },
              },
              "def": {
                "type": "string",
              },
              "~standard": {
                "validate": [Function],
                "vendor": "zod",
                "version": 1,
              },
            },
          ],
          "rest": null,
          "type": "tuple",
        },
      },
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "output": {
      "_zod": {
        "def": {
          "checks": [],
          "type": "number",
        },
      },
      "def": {
        "checks": [],
        "type": "number",
      },
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "type": "function",
  },
}
`;

exports[`snapshot > should z.function() match snapshot 2`] = `
$ZodFunction {
  "_def": {
    "input": {
      "_zod": {
        "def": {
          "items": [
            {
              "_zod": {
                "def": {
                  "type": "string",
                },
              },
              "def": {
                "type": "string",
              },
              "~standard": {
                "validate": [Function],
                "vendor": "zod",
                "version": 1,
              },
            },
          ],
          "rest": null,
          "type": "tuple",
        },
      },
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "output": {
      "_zod": {
        "def": {
          "checks": [],
          "type": "number",
        },
      },
      "def": {
        "checks": [],
        "type": "number",
      },
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "type": "function",
  },
  "def": {
    "input": {
      "_zod": {
        "def": {
          "items": [
            {
              "_zod": {
                "def": {
                  "type": "string",
                },
              },
              "def": {
                "type": "string",
              },
              "~standard": {
                "validate": [Function],
                "vendor": "zod",
                "version": 1,
              },
            },
          ],
          "rest": null,
          "type": "tuple",
        },
      },
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "output": {
      "_zod": {
        "def": {
          "checks": [],
          "type": "number",
        },
      },
      "def": {
        "checks": [],
        "type": "number",
      },
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "type": "function",
  },
}
`;

exports[`snapshot > should z.intersection() match snapshot 1`] = `
{
  "_zod": {
    "def": {
      "left": {
        "_zod": {
          "def": {
            "type": "string",
          },
        },
        "def": {
          "type": "string",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      "right": {
        "_zod": {
          "def": {
            "checks": [],
            "type": "number",
          },
        },
        "def": {
          "checks": [],
          "type": "number",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      "type": "intersection",
    },
  },
  "def": {
    "left": {
      "_zod": {
        "def": {
          "type": "string",
        },
      },
      "def": {
        "type": "string",
      },
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "right": {
      "_zod": {
        "def": {
          "checks": [],
          "type": "number",
        },
      },
      "def": {
        "checks": [],
        "type": "number",
      },
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "type": "intersection",
  },
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`snapshot > should z.intersection() match snapshot 2`] = `
{
  "_zod": {
    "def": {
      "left": {
        "_zod": {
          "def": {
            "type": "string",
          },
        },
        "def": {
          "type": "string",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      "right": {
        "_zod": {
          "def": {
            "checks": [],
            "type": "number",
          },
        },
        "def": {
          "checks": [],
          "type": "number",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      "type": "intersection",
    },
  },
  "def": {
    "left": {
      "_zod": {
        "def": {
          "type": "string",
        },
      },
      "def": {
        "type": "string",
      },
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "right": {
      "_zod": {
        "def": {
          "checks": [],
          "type": "number",
        },
      },
      "def": {
        "checks": [],
        "type": "number",
      },
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "type": "intersection",
  },
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`snapshot > should z.literal() match snapshot 1`] = `
{
  "_zod": {
    "def": {
      "type": "literal",
      "values": [
        "test",
      ],
    },
  },
  "def": {
    "type": "literal",
    "values": [
      "test",
    ],
  },
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`snapshot > should z.literal() match snapshot 2`] = `
{
  "_zod": {
    "def": {
      "type": "literal",
      "values": [
        "test",
      ],
    },
  },
  "def": {
    "type": "literal",
    "values": [
      "test",
    ],
  },
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`snapshot > should z.object() match snapshot 1`] = `
{
  "_zod": {
    "def": {
      "shape": {
        "bar": {
          "_zod": {
            "def": {
              "innerType": {
                "_zod": {
                  "def": {
                    "checks": [],
                    "type": "number",
                  },
                },
                "def": {
                  "checks": [],
                  "type": "number",
                },
                "~standard": {
                  "validate": [Function],
                  "vendor": "zod",
                  "version": 1,
                },
              },
              "type": "nullable",
            },
          },
          "def": {
            "innerType": {
              "_zod": {
                "def": {
                  "checks": [],
                  "type": "number",
                },
              },
              "def": {
                "checks": [],
                "type": "number",
              },
              "~standard": {
                "validate": [Function],
                "vendor": "zod",
                "version": 1,
              },
            },
            "type": "nullable",
          },
          "~standard": {
            "validate": [Function],
            "vendor": "zod",
            "version": 1,
          },
        },
        "foo": {
          "_zod": {
            "def": {
              "type": "string",
            },
          },
          "def": {
            "type": "string",
          },
          "~standard": {
            "validate": [Function],
            "vendor": "zod",
            "version": 1,
          },
        },
      },
      "type": "object",
    },
  },
  "def": {
    "shape": {
      "bar": {
        "_zod": {
          "def": {
            "innerType": {
              "_zod": {
                "def": {
                  "checks": [],
                  "type": "number",
                },
              },
              "def": {
                "checks": [],
                "type": "number",
              },
              "~standard": {
                "validate": [Function],
                "vendor": "zod",
                "version": 1,
              },
            },
            "type": "nullable",
          },
        },
        "def": {
          "innerType": {
            "_zod": {
              "def": {
                "checks": [],
                "type": "number",
              },
            },
            "def": {
              "checks": [],
              "type": "number",
            },
            "~standard": {
              "validate": [Function],
              "vendor": "zod",
              "version": 1,
            },
          },
          "type": "nullable",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      "foo": {
        "_zod": {
          "def": {
            "type": "string",
          },
        },
        "def": {
          "type": "string",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
    },
    "type": "object",
  },
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`snapshot > should z.object() match snapshot 2`] = `
{
  "_zod": {
    "def": {
      "shape": {
        "bar": {
          "_zod": {
            "def": {
              "innerType": {
                "_zod": {
                  "def": {
                    "checks": [],
                    "type": "number",
                  },
                },
                "def": {
                  "checks": [],
                  "type": "number",
                },
                "~standard": {
                  "validate": [Function],
                  "vendor": "zod",
                  "version": 1,
                },
              },
              "type": "nullable",
            },
          },
          "def": {
            "innerType": {
              "_zod": {
                "def": {
                  "checks": [],
                  "type": "number",
                },
              },
              "def": {
                "checks": [],
                "type": "number",
              },
              "~standard": {
                "validate": [Function],
                "vendor": "zod",
                "version": 1,
              },
            },
            "type": "nullable",
          },
          "~standard": {
            "validate": [Function],
            "vendor": "zod",
            "version": 1,
          },
        },
        "foo": {
          "_zod": {
            "def": {
              "type": "string",
            },
          },
          "def": {
            "type": "string",
          },
          "~standard": {
            "validate": [Function],
            "vendor": "zod",
            "version": 1,
          },
        },
      },
      "type": "object",
    },
  },
  "def": {
    "shape": {
      "bar": {
        "_zod": {
          "def": {
            "innerType": {
              "_zod": {
                "def": {
                  "checks": [],
                  "type": "number",
                },
              },
              "def": {
                "checks": [],
                "type": "number",
              },
              "~standard": {
                "validate": [Function],
                "vendor": "zod",
                "version": 1,
              },
            },
            "type": "nullable",
          },
        },
        "def": {
          "innerType": {
            "_zod": {
              "def": {
                "checks": [],
                "type": "number",
              },
            },
            "def": {
              "checks": [],
              "type": "number",
            },
            "~standard": {
              "validate": [Function],
              "vendor": "zod",
              "version": 1,
            },
          },
          "type": "nullable",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      "foo": {
        "_zod": {
          "def": {
            "type": "string",
          },
        },
        "def": {
          "type": "string",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
    },
    "type": "object",
  },
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`snapshot > should z.record() match snapshot 1`] = `
{
  "_zod": {
    "def": {
      "keyType": {
        "_zod": {
          "def": {
            "type": "string",
          },
        },
        "def": {
          "type": "string",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      "type": "record",
      "valueType": {
        "_zod": {
          "def": {
            "checks": [],
            "type": "number",
          },
        },
        "def": {
          "checks": [],
          "type": "number",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
    },
  },
  "def": {
    "keyType": {
      "_zod": {
        "def": {
          "type": "string",
        },
      },
      "def": {
        "type": "string",
      },
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "type": "record",
    "valueType": {
      "_zod": {
        "def": {
          "checks": [],
          "type": "number",
        },
      },
      "def": {
        "checks": [],
        "type": "number",
      },
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
  },
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`snapshot > should z.record() match snapshot 2`] = `
{
  "_zod": {
    "def": {
      "keyType": {
        "_zod": {
          "def": {
            "type": "string",
          },
        },
        "def": {
          "type": "string",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      "type": "record",
      "valueType": {
        "_zod": {
          "def": {
            "checks": [],
            "type": "number",
          },
        },
        "def": {
          "checks": [],
          "type": "number",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
    },
  },
  "def": {
    "keyType": {
      "_zod": {
        "def": {
          "type": "string",
        },
      },
      "def": {
        "type": "string",
      },
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "type": "record",
    "valueType": {
      "_zod": {
        "def": {
          "checks": [],
          "type": "number",
        },
      },
      "def": {
        "checks": [],
        "type": "number",
      },
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
  },
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`snapshot > should z.string() match snapshot 1`] = `
{
  "_zod": {
    "def": {
      "type": "string",
    },
  },
  "def": {
    "type": "string",
  },
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`snapshot > should z.string() match snapshot 2`] = `
{
  "_zod": {
    "def": {
      "type": "string",
    },
  },
  "def": {
    "type": "string",
  },
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`snapshot > should z.tuple() match snapshot 1`] = `
{
  "_zod": {
    "def": {
      "items": [
        {
          "_zod": {
            "def": {
              "type": "string",
            },
          },
          "def": {
            "type": "string",
          },
          "~standard": {
            "validate": [Function],
            "vendor": "zod",
            "version": 1,
          },
        },
        {
          "_zod": {
            "def": {
              "checks": [],
              "type": "number",
            },
          },
          "def": {
            "checks": [],
            "type": "number",
          },
          "~standard": {
            "validate": [Function],
            "vendor": "zod",
            "version": 1,
          },
        },
      ],
      "rest": {
        "_zod": {
          "def": {
            "type": "boolean",
          },
        },
        "def": {
          "type": "boolean",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      "type": "tuple",
    },
  },
  "def": {
    "items": [
      {
        "_zod": {
          "def": {
            "type": "string",
          },
        },
        "def": {
          "type": "string",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      {
        "_zod": {
          "def": {
            "checks": [],
            "type": "number",
          },
        },
        "def": {
          "checks": [],
          "type": "number",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
    ],
    "rest": {
      "_zod": {
        "def": {
          "type": "boolean",
        },
      },
      "def": {
        "type": "boolean",
      },
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "type": "tuple",
  },
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`snapshot > should z.tuple() match snapshot 2`] = `
{
  "_zod": {
    "def": {
      "items": [
        {
          "_zod": {
            "def": {
              "type": "string",
            },
          },
          "def": {
            "type": "string",
          },
          "~standard": {
            "validate": [Function],
            "vendor": "zod",
            "version": 1,
          },
        },
        {
          "_zod": {
            "def": {
              "checks": [],
              "type": "number",
            },
          },
          "def": {
            "checks": [],
            "type": "number",
          },
          "~standard": {
            "validate": [Function],
            "vendor": "zod",
            "version": 1,
          },
        },
      ],
      "rest": {
        "_zod": {
          "def": {
            "type": "boolean",
          },
        },
        "def": {
          "type": "boolean",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      "type": "tuple",
    },
  },
  "def": {
    "items": [
      {
        "_zod": {
          "def": {
            "type": "string",
          },
        },
        "def": {
          "type": "string",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      {
        "_zod": {
          "def": {
            "checks": [],
            "type": "number",
          },
        },
        "def": {
          "checks": [],
          "type": "number",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
    ],
    "rest": {
      "_zod": {
        "def": {
          "type": "boolean",
        },
      },
      "def": {
        "type": "boolean",
      },
      "~standard": {
        "validate": [Function],
        "vendor": "zod",
        "version": 1,
      },
    },
    "type": "tuple",
  },
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`snapshot > should z.union() match snapshot 1`] = `
{
  "_zod": {
    "def": {
      "options": [
        {
          "_zod": {
            "def": {
              "type": "string",
            },
          },
          "def": {
            "type": "string",
          },
          "~standard": {
            "validate": [Function],
            "vendor": "zod",
            "version": 1,
          },
        },
        {
          "_zod": {
            "def": {
              "checks": [],
              "type": "number",
            },
          },
          "def": {
            "checks": [],
            "type": "number",
          },
          "~standard": {
            "validate": [Function],
            "vendor": "zod",
            "version": 1,
          },
        },
      ],
      "type": "union",
    },
  },
  "def": {
    "options": [
      {
        "_zod": {
          "def": {
            "type": "string",
          },
        },
        "def": {
          "type": "string",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      {
        "_zod": {
          "def": {
            "checks": [],
            "type": "number",
          },
        },
        "def": {
          "checks": [],
          "type": "number",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
    ],
    "type": "union",
  },
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;

exports[`snapshot > should z.union() match snapshot 2`] = `
{
  "_zod": {
    "def": {
      "options": [
        {
          "_zod": {
            "def": {
              "type": "string",
            },
          },
          "def": {
            "type": "string",
          },
          "~standard": {
            "validate": [Function],
            "vendor": "zod",
            "version": 1,
          },
        },
        {
          "_zod": {
            "def": {
              "checks": [],
              "type": "number",
            },
          },
          "def": {
            "checks": [],
            "type": "number",
          },
          "~standard": {
            "validate": [Function],
            "vendor": "zod",
            "version": 1,
          },
        },
      ],
      "type": "union",
    },
  },
  "def": {
    "options": [
      {
        "_zod": {
          "def": {
            "type": "string",
          },
        },
        "def": {
          "type": "string",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      {
        "_zod": {
          "def": {
            "checks": [],
            "type": "number",
          },
        },
        "def": {
          "checks": [],
          "type": "number",
        },
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
    ],
    "type": "union",
  },
  "~standard": {
    "validate": [Function],
    "vendor": "zod",
    "version": 1,
  },
}
`;
